subsys/net/lib/lwm2m/lwm2m_registry.c

```diff
@@ -1344,3 +1354,252 @@ bool lwm2m_engine_shall_report_obj_version(const struct lwm2m_engine_obj *obj)
+struct lwm2m_time_series_resource *lwm2m_cache_entry_get_by_string(char const *resource_path)
+{
+#if defined(CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT)
+	if (!sys_slist_is_empty(&lwm2m_timed_cache_list)) {
+		int ret;
+		struct lwm2m_time_series_resource *entry;
+
+		/* Keep list Alphabetical order */
+		SYS_SLIST_FOR_EACH_CONTAINER(&lwm2m_timed_cache_list, entry, node) {
+			ret = strcmp(entry->path, resource_path);
+			if (ret == 0) {
+				return entry;
+			} else if (ret > 0) {
+				return NULL;
+			}
+		}
+	}
+#endif /* CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT */
+	return NULL;
+}
+
+struct lwm2m_time_series_resource *lwm2m_cache_entry_get_by_object(struct lwm2m_obj_path *obj_path)
+{
+#if defined(CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT)
+	char obj_path_str[25];
+	char const *resource_path;
+
+	if (!obj_path || obj_path->level < LWM2M_PATH_LEVEL_RESOURCE) {
+		LOG_ERR("Path level wrong for cache %u", obj_path->level);
+		return NULL;
+	}
+
+	/* Decode path to string */
+	resource_path = lwm2m_path_log_buf(obj_path_str, obj_path);
+
+	return lwm2m_cache_entry_get_by_string(resource_path);
+#else
+	return NULL;
+#endif /* CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT */
+}
+
+int lwm2m_engine_enable_cache(char const *resource_path, struct lwm2m_time_series_elem *data_cache,
+			      size_t cache_len)
+{
+#if defined(CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT)
+	struct lwm2m_obj_path path;
+	struct lwm2m_engine_obj_inst *obj_inst;
+	struct lwm2m_engine_obj_field *obj_field;
+	struct lwm2m_engine_res_inst *res_inst = NULL;
+	struct lwm2m_time_series_resource *cache_entry;
+	int ret = 0;
+
+	/* translate path -> path_obj */
+	ret = lwm2m_string_to_path(resource_path, &path, '/');
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (path.level < 3) {
+		LOG_ERR("path must have at least 3 parts");
+		return -EINVAL;
+	}
+
+	/* look up resource obj */
+	ret = path_to_objs(&path, &obj_inst, &obj_field, NULL, &res_inst);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (!res_inst) {
+		LOG_ERR("res instance %d not found", path.res_inst_id);
+		return -ENOENT;
+	}
+
+	switch (obj_field->data_type) {
+	case LWM2M_RES_TYPE_U32:
+	case LWM2M_RES_TYPE_TIME:
+	case LWM2M_RES_TYPE_U16:
+	case LWM2M_RES_TYPE_U8:
+	case LWM2M_RES_TYPE_S64:
+	case LWM2M_RES_TYPE_S32:
+	case LWM2M_RES_TYPE_S16:
+	case LWM2M_RES_TYPE_S8:
+	case LWM2M_RES_TYPE_BOOL:
+	case LWM2M_RES_TYPE_FLOAT:
+		/* Support only fixed width resource types */
+		cache_entry = lwm2m_cache_entry_allocate(resource_path);
+		break;
+	default:
+		cache_entry = NULL;
+		break;
+	}
+
+	if (!cache_entry) {
+		return -ENODATA;
+	}
+
+	cache_entry->capacity = cache_len;
+	cache_entry->size = 0;
+	cache_entry->elements = data_cache;
+	return 0;
+#else
+	LOG_ERR("LwM2M resource cache is only supported for "
+		"CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT");
+	return -ENOTSUP;
+#endif /* CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT */
+}
+
+int lwm2m_engine_data_cache_init(void)
+{
+#if defined(CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT)
```

Comment:
I would assume this check can be removed as the calller checks it by using _IS_ENABLED(CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT)_.